File: platform-services/backstage/templates/_fragments/backend/rails/skeleton/test/integration/health_test.rb
require "test_helper"

# Health + secret-proof endpoints (no database needed).
class HealthTest < ActionDispatch::IntegrationTest
  test "GET /healthz is 200 and DB-independent" do
    get "/healthz"
    assert_response :success
    assert_equal({ "status" => "ok" }, JSON.parse(response.body))
  end

  test "GET /health is 200" do
    get "/health"
    assert_response :success
    assert_equal({ "status" => "ok" }, JSON.parse(response.body))
  end

  test "GET / proves APP_SECRET was read without echoing it" do
    ENV["APP_SECRET"] = "hunter2"
    get "/"
    assert_response :success
    body = JSON.parse(response.body)
    assert_equal true, body["secret_loaded"]
    assert_equal "hunter2".length, body["secret_length"]
    refute_includes response.body, "hunter2"
  ensure
    ENV.delete("APP_SECRET")
  end

  test "GET / reports secret missing when APP_SECRET is unset" do
    ENV.delete("APP_SECRET")
    get "/"
    assert_response :success
    body = JSON.parse(response.body)
    assert_equal false, body["secret_loaded"]
    assert_equal 0, body["secret_length"]
  end
end

File: platform-services/backstage/templates/_fragments/backend/laravel/skeleton/tests/Feature/HealthTest.php
<?php

namespace Tests\Feature;

use Tests\TestCase;

// Health + secret-proof endpoints (no database needed).
class HealthTest extends TestCase
{
    public function test_healthz_is_ok_and_db_independent(): void
    {
        $this->getJson('/healthz')->assertOk()->assertExactJson(['status' => 'ok']);
    }

    public function test_health_alias_is_ok(): void
    {
        $this->getJson('/health')->assertOk()->assertExactJson(['status' => 'ok']);
    }

    public function test_root_proves_secret_without_echoing_it(): void
    {
        putenv('APP_SECRET=hunter2');
        $_ENV['APP_SECRET'] = 'hunter2';
        $response = $this->getJson('/');
        $response->assertOk()
            ->assertJson(['secret_loaded' => true, 'secret_length' => 7]);
        $this->assertStringNotContainsString('hunter2', $response->getContent());
        putenv('APP_SECRET');
        unset($_ENV['APP_SECRET']);
    }

    public function test_root_reports_secret_missing_when_unset(): void
    {
        putenv('APP_SECRET');
        unset($_ENV['APP_SECRET']);
        $this->getJson('/')
            ->assertOk()
            ->assertJson(['secret_loaded' => false, 'secret_length' => 0]);
    }

    public function test_data_routes_degrade_to_503_when_db_unconfigured(): void
    {
        // Mirror the production "DATABASE_URL unset" path: default = mysql, no url. No
        // RefreshDatabase here, so nothing tries to connect to that (absent) MySQL — the
        // controller's guard returns 503 before touching the DB.
        config(['database.default' => 'mysql', 'database.connections.mysql.url' => null]);
        $this->getJson('/api/items')
            ->assertStatus(503)
            ->assertJson(['error' => 'database unavailable: DATABASE_URL is not set or the database is unreachable']);
    }
}
